core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1637,6 +1640,83 @@ public void sync(Translog.Location location) {
         }
     }
 
+    private final ArrayBlockingQueue<Tuple<Translog.Location, Consumer<Exception>>> fsyncQueue = new ArrayBlockingQueue<>(1024);
+    private final Semaphore promiseSync = new Semaphore(1);
+
+    /**
+     * Syncs the given location with the underlying storage unless already synced. This method might return immediately without
+     * actually fsyncing the location until the sync listener is called. Yet, unless there is already another thread fsyncing
+     * the transaction log the caller thread will be hijacked to run the fsync for all pending fsync operations.
+     * This method allows indexing threads to continue indexing without blocking on fsync calls. We ensure that there is only
+     * one thread blocking on the sync an all others can continue indexing.
+     */
+    public final void sync(Translog.Location location, Consumer<Exception> syncListener) {
+        verifyNotClosed();
+        // the algorithm here tires to reduce the load on each individual caller.
+        // we try to have only one caller that syncs pending locations to disc while others just add to the queue but
+        // at the same time never overload the node by pushing too many items into the queue.
+
+        // we first try make a promise that we are responsible for the sync
+        final boolean promised = promiseSync.tryAcquire();
+        final List<Tuple<Translog.Location, Consumer<Exception>>> candidates = new ArrayList<>();
+        if (promised) {
+            // if we are responsible we can't wait for others to make space - we have to sync until we can add to the queue
+            while (fsyncQueue.offer(new Tuple<>(location, syncListener)) == false) {
+                drainAndSync(candidates);
+            }
+        } else {
+            // in this case we are not responsible and can just block until there is space
+            try {
+                fsyncQueue.put(new Tuple<>(location, syncListener));
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+
+        // here we have to try to make the promise again otherwise there is a race when a thread puts an entry without making the promise
+        // while we are draining that mean we might exit below too early in the while loop if the drainAndSync call is fast.
+        if (promised || promiseSync.tryAcquire()) {
+            // since we made the promise to sync we gotta do it here at least once
+            drainAndSync(candidates);
+            promiseSync.release(); // now to ensure we are passing it on we release the promise so another thread can take over
```

Comment:
this would mean that there are more than one thread fsyncing which is against what I am trying to achieve here. The point is that we try to batch up incoming locations while we fsync to not block too many threads and let them do the indexing. Makes sense? I am happy to add a comment